### ISSUE TITLE
“Link abstraction” ignores “Should page URLs end with a slash” option

### ISSUE THREAD

**Issue by kondurake** [processwire/processwire-issues#286]
### Short description of the issue
Enabling “Link abstraction” adds a trailing slash for relative URLs even if option “Should page URLs end with a slash” is disabled.

For example, I have a page “/my-page” and I disabled trailing slashes for its template (i.e, “Should page URLs end with a slash” is set to “No”). If I save another page and paste the link below into a textarea that has enabled “Link abstraction”:
`<a href="/my-page">My Page</a>`
it is converted to:
`<a href="/my-page/">My Page</a>`

### Expected behavior
When I use relative URLs and option “Should page URLs end with a slash” is equal to “No” it should not add a trailing slash.

### Actual behavior
It ignores “Should page URLs end with a slash” option and adds a trailing slash.

### Suggestion for a possible fix
Open file **/wire/core/MarkupQA.php** and find following code (at line ~425):
```
$livePath = $this->wire('pages')->getPath((int) $pageID, array(
	'language' => $language
));
```

After it add the following code:
```
if (!$this->page->template->slashUrls) {
	$livePath = rtrim($livePath, '/');
}
```
Unfortunately, I don’t know if this workaround can have a negative impact on other functions. On other hand, the right place for fixing it should be the `PagesLoader->getPath` function, but I suppose it can break even more things. And the last one, please note that I’m using Nginx and cannot confirm if this _bug_ occurs on other environments.

### Setup/Environment
* ProcessWire version: 3.0.62
* PHP version: 5.5.0
* MySQL version: 5.5.5
* Nginx: 1.13.0

**Comment by ryancramerdesign** [comment]
Thanks, I've pushed a fix for this issue. A page "path" is always inclusive of trailing slashes, since that's how PW internally refers to all pages. So the `$pages->getPath()` method is returning what it should. But the MarkupQA is what needs to retain the lack of trailing slash. The solution you mentioned won't work because `$this->page` is referring to the page that is having the field rendered, not the page being linked to. And actually the page being linked to is typically not loaded at all, as a matter of efficiency. So we can't query its template to check for a trailing slash setting. But what we can do is retain whatever the user specified in their link. So if you specified no trailing slash in your link, that should now be retained (as of the referenced commit). Thanks.

**Comment by kondurake** [comment]
Hi Ryan! Thank you for explanations. Even I’m new to PW, sorry for the wrong “fix” — next time I’ll try to be carefully.

**Comment by LostKobrakai** [comment]
> Even I’m new to PW, sorry for the wrong “fix” — next time I’ll try to be carefully.

I don't think this was an accusation at all. It's nice you even tried to fix it.

**Comment by ryancramerdesign** [comment]
> sorry for the wrong “fix” — next time I’ll try to be carefully.

> I don't think this was an accusation at all. It's nice you even tried to fix it.

I agree. I always appreciate it when people suggest fixes because it actually helps a lot in understanding and fixing the issue, regardless of whether the suggested fix works or not. Keep doing what you are doing. 👍 Thanks.

**Comment by kondurake** [comment]
Of course, I did not mean that Ryan accused me. I’m sorry because of my actions he had to explain some obvious things.

By the way, thank you both for your kind words.

**Comment by netcarver** [comment]
@kondurake Hello Viktor. Could you close this issue if it is resolved to your satisfaction, otherwise post a reply so we know to keep it open and on the radar. Thank you!
